utils/pennsylvania.py

```diff
@@ -0,0 +1,643 @@
+        donor_organizations = donor_organizations.rename(
+            columns={"DONOR_ID": "id", "DONOR": "name", "DONOR_TYPE": "entity_type"}
+        )
+        recipient_organizations = organizations_df.loc[
+            (
+                (organizations_df.RECIPIENT_TYPE == "Committee")
+                | (organizations_df.RECIPIENT_TYPE == "Organization")
+            )
+        ]
+        recipient_organizations = recipient_organizations[
+            ["RECIPIENT_ID", "RECIPIENT", "RECIPIENT_TYPE"]
+        ]
+        recipient_organizations = recipient_organizations.rename(
+            columns={
+                "RECIPIENT_ID": "id",
+                "RECIPIENT": "name",
+                "RECIPIENT_TYPE": "entity_type",
+            }
+        )
+        all_organizations = pd.concat([donor_organizations, recipient_organizations])
+        all_organizations = all_organizations.drop_duplicates()
+        all_organizations["state"] = "PA"
+
+        return all_organizations
+
+    def make_transactions_table(organizations_df: pd.DataFrame) -> pd.DataFrame:
+        """This function creates a transaction_type column describing the
+        type of contribution from a donor to a recipient, and assigns the
+        appropriate office for which the contribution was made for.
+        Args:
+            df: a pandas dataframe with donor and recipient information that
+            details relevant information about a singular transactions,
+            including a transaction ID, donor and recipient IDs, and the
+            donation amount
+        Returns:
+            a pandas dataframe strictly with information regarding transactions
+            between donors and recipients.
+        """
+
+        df = organizations_df.copy()
+        column_names = list(organizations_df.columns.str.lower())
+        df.columns = column_names
+        df["transaction_type"] = None
+        df["office_sought"] = df.apply(
+            lambda x: x["donor_office"]
+            if x["recipient_office"] is None
+            else x["recipient_office"],
+            axis=1,
+        )
+        df = df.drop(columns={"donor_office", "recipient_office"})
+
+        return df
+
+    def preprocess(self, filepaths_list: list[str]) -> list[pd.DataFrame]:
+
+        contributor_datasets, filer_datasets, expense_datasets = [], [], []
+        # incase the provided filepath was not in increasing order, this just
+        # makes sure the filepaths numerically line up
+        filepaths_list.sort()
+        for path in filepaths_list:
+            # because of the different formatting for data in different years
+            # obstained from the PA website years, I had my webscraper append
+            # the year to the filename
+            dir = path.split("/")
+            file_name = dir[len(dir) - 1]
+
+            # only want contributor, filer, and expenditure files:
+            if (
```

Comment:
@alankagiri is there a situation where this is not true?

Reply:
the webscraper collects all files...contributor, filer, expenditure, debt, and receipts. So I only want to bother executing the code when the file name indicates any of the desired 3